drivers/video/video_ctrls.c

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2025 NXP
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/drivers/video.h>
+#include <zephyr/drivers/video-controls.h>
+#include <zephyr/logging/log.h>
+
+#include "video_ctrls.h"
+#include "video_device.h"
+
+LOG_MODULE_REGISTER(video_ctrls, CONFIG_VIDEO_LOG_LEVEL);
+
+void video_ctrl_handler_init(struct video_ctrl_handler *hdl, const struct device *dev)
+{
+	struct video_device *vdev = video_device_find(dev);
+
+	if (!hdl || !vdev) {
+		return;
+	}
+
+	vdev->ctrl_handler = hdl;
+	sys_slist_init(&hdl->list);
+}
+
+static inline int check_range(enum video_ctrl_type type, int32_t min, int32_t max, uint32_t step,
+			      int32_t def)
+{
+	switch (type) {
+	case VIDEO_CTRL_TYPE_BOOLEAN:
+		if (step != 1 || max > 1 || min < 0) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_INTEGER:
+	case VIDEO_CTRL_TYPE_INTEGER64:
+		if (step == 0 || min > max || !IN_RANGE(def, min, max)) {
+			return -ERANGE;
+		}
+		return 0;
+	case VIDEO_CTRL_TYPE_MENU:
+	case VIDEO_CTRL_TYPE_INTEGER_MENU:
+		if (!IN_RANGE(min, 0, max) || !IN_RANGE(def, min, max)) {
+			return -ERANGE;
+		}
+		return 0;
+	default:
+		return 0;
+	}
+}
+
+static inline void set_type_flag(uint32_t id, enum video_ctrl_type *type, uint32_t *flags)
+{
+	*flags = 0;
+
+	switch (id) {
+	case VIDEO_CID_HFLIP:
+	case VIDEO_CID_VFLIP:
+		*type = VIDEO_CTRL_TYPE_BOOLEAN;
+		break;
+	case VIDEO_CID_POWER_LINE_FREQUENCY:
+	case VIDEO_CID_TEST_PATTERN:
+		*type = VIDEO_CTRL_TYPE_MENU;
+		break;
+	case VIDEO_CID_PIXEL_RATE:
+		*type = VIDEO_CTRL_TYPE_INTEGER64;
+		*flags |= VIDEO_CTRL_FLAG_READ_ONLY;
+		break;
+	default:
+		*type = VIDEO_CTRL_TYPE_INTEGER;
+		break;
+	}
+}
+
+int video_ctrl_init(struct video_ctrl *ctrl, struct video_ctrl_handler *hdl, uint32_t id,
+		    int32_t min, int32_t max, uint32_t step, int32_t def)
+{
+	int ret;
+	enum video_ctrl_type type;
+	uint32_t flags;
+
+	set_type_flag(id, &type, &flags);
+
+	/* Sanity checks */
+	if (id == 0 || id >= VIDEO_CID_PRIVATE_BASE) {
+		return -EINVAL;
+	}
+
+	ret = check_range(type, min, max, step, def);
+	if (ret) {
+		return ret;
+	}
+
+	ctrl->handler = hdl;
+
+	ctrl->id = id;
+	ctrl->type = type;
+	ctrl->flags = flags;
+	ctrl->min = min;
+	ctrl->max = max;
+	ctrl->step = step;
+	ctrl->def = def;
+	ctrl->val = def;
+
+	sys_slist_append(&hdl->list, &ctrl->node);
+
+	return 0;
+}
+
+struct video_ctrl *video_ctrl_find(struct video_ctrl_handler *hdl, uint32_t id)
+{
+	struct video_ctrl *ctrl;
+
+	if (!hdl) {
+		return NULL;
+	}
+
+	SYS_SLIST_FOR_EACH_CONTAINER(&hdl->list, ctrl, node) {
+		if (ctrl->id == id) {
+			return ctrl;
+		}
+	}
+
+	return NULL;
+}
+
+static int video_ctrl_find_by_dev(const struct device *dev, uint32_t id, struct video_ctrl **ctrl)
```

Comment:
```suggestion
static int video_ctrl_find_recursive(const struct device *dev, uint32_t id, struct video_ctrl **ctrl)
```

This is not important, though `_find_by_dev()` first made me think that it was only searching controls within a single device, while it is searching beyond this device down onto the list of child devices.

Reply:
Yes, now we only have `video_find_ctrl()`